pkg/controller/registryinfra/registryinfra_controller.go

```diff
@@ -0,0 +1,696 @@
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not an Object %#v", obj))
+			return
+		}
+	}
+	c.logger.Debugf("enqueueing deleted cluster %s/%s", cluster.GetNamespace(), cluster.GetName())
+	c.enqueueCluster(cluster)
+}
+
+// Run runs c; will not return until stopCh is closed. workers determines how
+// many clusters will be handled in parallel.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Infof("starting %s controller", controllerName)
+	defer c.logger.Infof("shutting down %s controller", controllerName)
+
+	if !controller.WaitForCacheSync("registryinfra", stopCh, c.clustersSynced) {
+		c.logger.Errorf("Could not sync caches for %s controller", controllerName)
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(cluster metav1.Object) {
+	key, err := controller.KeyFunc(cluster)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", cluster, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	logger := c.logger.WithField("cluster", key)
+
+	logger.Errorf("error syncing registry infra: %v", err)
+	if c.queue.NumRequeues(key) < maxRetries {
+		logger.Errorf("retrying registry infra")
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("dropping cluster out of the queue: %v", err)
+	c.queue.Forget(key)
+}
+
+func (c *Controller) syncRegistryInfra(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing registry infra")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing registry infra")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	capicluster, err := c.clusterLister.Clusters(namespace).Get(name)
+	if err != nil {
+		return fmt.Errorf("error retrieving cluster: %v", err)
+	}
+
+	cluster, err := controller.ConvertToCombinedCluster(capicluster)
+	if err != nil {
+		return fmt.Errorf("failure converting to combined cluster: %v", err)
+	}
+
+	clusterDeploymentStatus, err := controller.ClusterProviderStatusFromCluster(capicluster)
```

Comment:
I think this is clusterStatus, or providerStatus, and separate from clusterDeploymentStatus.